<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Main function to perform cell type deconvolution with MeDIP-seq data — decemedip • decemedip</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Main function to perform cell type deconvolution with MeDIP-seq data — decemedip"><meta name="description" content="Main function to perform cell type deconvolution with MeDIP-seq data"><meta property="og:description" content="Main function to perform cell type deconvolution with MeDIP-seq data"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">decemedip</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/how-to-use-decemedip.html">Cell type deconvolutiond of (cf)MeDIP-seq data with decemedip</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Main function to perform cell type deconvolution with MeDIP-seq data</h1>

      <div class="d-none name"><code>decemedip.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Main function to perform cell type deconvolution with MeDIP-seq data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">decemedip</span><span class="op">(</span></span>
<span>  sample_bam_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  paired_end <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  counts_cts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  counts_anc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  ref_assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  ref_cts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ref_anc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  weight_cts <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  weight_anc <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  diagnostics <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">2024</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  chains <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  iter <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  stan_input_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>s_mu <span class="op">=</span> <span class="fl">3</span>, s_sigma <span class="op">=</span> <span class="fl">3</span>, n_knot_z <span class="op">=</span> <span class="fl">0</span>, degree_z <span class="op">=</span> <span class="fl">3</span>, Xi <span class="op">=</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">ref_cts</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, s_theta <span class="op">=</span> <span class="fl">3</span>, s_tau <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>,</span>
<span>  stan_control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>adapt_delta <span class="op">=</span> <span class="fl">0.95</span>, max_treedepth <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>,</span>
<span>  timeout_sec <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="op">(</span><span class="va">diagnostics</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="va">iter</span> <span class="op">*</span> <span class="va">chains</span>,</span>
<span>  max_retries <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sample-bam-file">sample_bam_file<a class="anchor" aria-label="anchor" href="#arg-sample-bam-file"></a></dt>
<dd><p>A string value that indicates the file path to the bam file of a
MeDIP-seq sample of interest. If <code>sample_bam_file</code> is specified, please do not
specify <code>counts_cts</code> and <code>counts_anc</code> to avoid conflict.</p></dd>


<dt id="arg-paired-end">paired_end<a class="anchor" aria-label="anchor" href="#arg-paired-end"></a></dt>
<dd><p>A logic value that indicates whether the bam file is from paired-end
reads or single-end. Specify <code>TRUE</code> for paired-end and <code>FALSE</code> for single-end.</p></dd>


<dt id="arg-counts-cts">counts_cts<a class="anchor" aria-label="anchor" href="#arg-counts-cts"></a></dt>
<dd><p>An atomic vector of integer values that indicates the read counts of
a MeDIP-seq sample on reference sites/regions. If <code>counts_cts</code> and <code>counts_anc</code>
is specified, please do not specify <code>sample_bam_file</code> and <code>paired_end</code> to avoid conflict.</p></dd>


<dt id="arg-counts-anc">counts_anc<a class="anchor" aria-label="anchor" href="#arg-counts-anc"></a></dt>
<dd><p>An atomic vector of integer values that indicates the read counts of
a MeDIP-seq sample on reference sites/regions. If  <code>counts_cts</code> and <code>counts_anc</code>
is specified, please do not specify <code>sample_bam_file</code> and <code>paired_end</code> to avoid conflict.</p></dd>


<dt id="arg-ref-assembly">ref_assembly<a class="anchor" aria-label="anchor" href="#arg-ref-assembly"></a></dt>
<dd><p>A string that represents the genome assembly that should be used for cell type-specific
sites in the reference panel ('hg19' or 'hg38'). Default to 'hg19'. The default reference is explained
in the manuscript of <span class="pkg">decemedip</span>. Alternatively, if the user want to provide their own
reference panel by using the <code>ref_cts</code> and <code>ref_anc</code> arguments.</p></dd>


<dt id="arg-ref-cts">ref_cts<a class="anchor" aria-label="anchor" href="#arg-ref-cts"></a></dt>
<dd><p>A <code>SummarizedExperiment</code> object that contains the genomic coordinates and
beta values of the cell type-specific sites/regions from reference cell types.
The <code><a href="makeReferencePanel.html">makeReferencePanel</a></code> can be used to generate such a panel.</p></dd>


<dt id="arg-ref-anc">ref_anc<a class="anchor" aria-label="anchor" href="#arg-ref-anc"></a></dt>
<dd><p>Same as <code>ref_cts</code> but for anchor sites.</p></dd>


<dt id="arg-weight-cts">weight_cts<a class="anchor" aria-label="anchor" href="#arg-weight-cts"></a></dt>
<dd><p>A numeric value indicating the weights that should be put on cell type-specific
sites/regions. Default is 0.5.</p></dd>


<dt id="arg-weight-anc">weight_anc<a class="anchor" aria-label="anchor" href="#arg-weight-anc"></a></dt>
<dd><p>A numeric value indicating the weights that should be put on cell type-specific
sites/regions. Default is 1.</p></dd>


<dt id="arg-diagnostics">diagnostics<a class="anchor" aria-label="anchor" href="#arg-diagnostics"></a></dt>
<dd><p>A logic value that indicates whether to include components of the stan model
in the output that are necessary for future diagnostics of the model, such as posterior
predictive checks. For details, please refer to the function <code><a href="plotDiagnostics.html">plotDiagnostics</a></code>.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>The seed for random number generation in MCMC sampling.</p></dd>


<dt id="arg-cores">cores<a class="anchor" aria-label="anchor" href="#arg-cores"></a></dt>
<dd><p>A positive integer specifying the number of cores that can be used for
MCMC sampling. The default is 4</p></dd>


<dt id="arg-chains">chains<a class="anchor" aria-label="anchor" href="#arg-chains"></a></dt>
<dd><p>A positive integer specifying the number of Markov chains. The default is 4.</p></dd>


<dt id="arg-iter">iter<a class="anchor" aria-label="anchor" href="#arg-iter"></a></dt>
<dd><p>A positive integer specifying the number of iterations for each chain (including
warmup). The default is 2000.</p></dd>


<dt id="arg-stan-input-params">stan_input_params<a class="anchor" aria-label="anchor" href="#arg-stan-input-params"></a></dt>
<dd><p>A named list of parameters that specifies the prior of the
decemedip model.</p></dd>


<dt id="arg-stan-control">stan_control<a class="anchor" aria-label="anchor" href="#arg-stan-control"></a></dt>
<dd><p>A named list of parameters to control the sampler's behavior in
Stan. See the details in the documentation for the control argument in <code><a href="https://mc-stan.org/rstan/reference/stan.html" class="external-link">stan</a></code>.</p></dd>


<dt id="arg-timeout-sec">timeout_sec<a class="anchor" aria-label="anchor" href="#arg-timeout-sec"></a></dt>
<dd><p>A numerical value indicating the CPU/processor time (in seconds) allowed for
the MCMC to run before restarting the chains with a new random seed.</p></dd>


<dt id="arg-max-retries">max_retries<a class="anchor" aria-label="anchor" href="#arg-max-retries"></a></dt>
<dd><p>An integer value indicating the maximum number of tries with different seed
for MCMC if it fails to converge.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other parameters that can be passed to the <code><a href="https://mc-stan.org/rstan/reference/stanmodel-method-sampling.html" class="external-link">sampling</a></code> function.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of two elements:</p><ol><li><p><code>data_list</code>: An organized list of variables used as input to the Stan posterior sampling function.</p></li>
<li><p><code>posterior</code>: An <code>stanfit</code> object produced by Stan representing the fitted posteriors.</p></li>
</ol></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pdx.counts.cts.se</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pdx.counts.anc.se</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># read counts of cell type-specific CpGs of the sample 'LuCaP_147CR'</span></span></span>
<span class="r-in"><span><span class="va">counts_cts</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">pdx.counts.cts.se</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>,<span class="st">'LuCaP_147CR'</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="co"># read counts of anchor CpGs of the sample 'LuCaP_147CR'</span></span></span>
<span class="r-in"><span><span class="va">counts_anc</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">pdx.counts.anc.se</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>,<span class="st">'LuCaP_147CR'</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="co"># Fit decemedip model (iter=100 for demonstration, by default iter=2000)</span></span></span>
<span class="r-in"><span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu">decemedip</span><span class="op">(</span>counts_cts <span class="op">=</span> <span class="va">counts_cts</span>, counts_anc <span class="op">=</span> <span class="va">counts_anc</span>, iter <span class="op">=</span> <span class="fl">100</span>, cores <span class="op">=</span> <span class="fl">1</span>, chains <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SAMPLING FOR MODEL 'decemedip1' NOW (CHAIN 1).</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Gradient evaluation took 0.00179 seconds</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 17.9 seconds.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Adjust your expectations accordingly!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: WARNING: There aren't enough warmup iterations to fit the</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:          three stages of adaptation as currently configured.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:          Reducing each adaptation stage to 15%/75%/10% of</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:          the given number of warmup iterations:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:            init_buffer = 7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:            adapt_window = 38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:            term_buffer = 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration:  1 / 100 [  1%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 10 / 100 [ 10%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 20 / 100 [ 20%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 30 / 100 [ 30%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 40 / 100 [ 40%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 50 / 100 [ 50%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 51 / 100 [ 51%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 60 / 100 [ 60%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 70 / 100 [ 70%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 80 / 100 [ 80%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 90 / 100 [ 90%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 100 / 100 [100%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:  Elapsed Time: 33.471 seconds (Warm-up)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:                22.133 seconds (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:                55.604 seconds (Total)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The largest R-hat is NA, indicating chains have not mixed.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Running the chains for more iterations may help. See</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> https://mc-stan.org/misc/warnings.html#r-hat</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Bulk Effective Samples Size (ESS) is too low, indicating posterior means and medians may be unreliable.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Running the chains for more iterations may help. See</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> https://mc-stan.org/misc/warnings.html#bulk-ess</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Tail Effective Samples Size (ESS) is too low, indicating posterior variances and tail quantiles may be unreliable.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Running the chains for more iterations may help. See</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> https://mc-stan.org/misc/warnings.html#tail-ess</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MCMC converged with seed 2024</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ning Shen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

