<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Obtain read counts of regions of interest from MeDIP-seq bam files — getRoiReadCount • decemedip</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Obtain read counts of regions of interest from MeDIP-seq bam files — getRoiReadCount"><meta name="description" content="Obtain read counts of regions of interest from MeDIP-seq bam files"><meta property="og:description" content="Obtain read counts of regions of interest from MeDIP-seq bam files"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">decemedip</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/how-to-use-decemedip.html">Cell type deconvolutiond of (cf)MeDIP-seq data with decemedip</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Obtain read counts of regions of interest from MeDIP-seq bam files</h1>

      <div class="d-none name"><code>getRoiReadCount.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Obtain read counts of regions of interest from MeDIP-seq bam files</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getRoiReadCount</span><span class="op">(</span></span>
<span>  <span class="va">sample_bam_files</span>,</span>
<span>  <span class="va">sample_names</span>,</span>
<span>  <span class="va">sample_paired</span>,</span>
<span>  <span class="va">roi</span>,</span>
<span>  col_data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  row_data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bs_genome <span class="op">=</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sample-bam-files">sample_bam_files<a class="anchor" aria-label="anchor" href="#arg-sample-bam-files"></a></dt>
<dd><p>An atomic vector that contains directory paths of the MeDIP-seq
sorted bam files.</p></dd>


<dt id="arg-sample-names">sample_names<a class="anchor" aria-label="anchor" href="#arg-sample-names"></a></dt>
<dd><p>An atomic vector of strings that indicates the sample names.
Default is NULL. If not NULL, please make sure that <code>sample_names</code> corresponds to
elements of <code>sample_bam_files</code>.</p></dd>


<dt id="arg-sample-paired">sample_paired<a class="anchor" aria-label="anchor" href="#arg-sample-paired"></a></dt>
<dd><p>A logic value that indicates sample_paired-end reads (if TRUE) or single-end
reads (if FALSE).</p></dd>


<dt id="arg-roi">roi<a class="anchor" aria-label="anchor" href="#arg-roi"></a></dt>
<dd><p>A <code>GRanges</code> object that contains the genomic coordinates of the region
of interest (ROI).</p></dd>


<dt id="arg-col-data">col_data<a class="anchor" aria-label="anchor" href="#arg-col-data"></a></dt>
<dd><p>A <code>DataFrame</code> object that contains metadata for columns (i.e.,
samples) if specified. Default is NULL. If not NULL, please make sure that
rows of <code>col_data</code> corresponds to elements of sample_bam_files. If input is a
non-<code>DataFrame</code> object, it will be converted to a <code>DataFrame</code>.</p></dd>


<dt id="arg-row-data">row_data<a class="anchor" aria-label="anchor" href="#arg-row-data"></a></dt>
<dd><p>A <code>DataFrame</code> object that contains metadata for rows (i.e.,
genomic regions) if specified. Default is NULL. If not NULL, please make sure that
rows of <code>row_data</code> corresponds to elements of sample_bam_files. If input is a
non-<code>DataFrame</code> object, it will be converted to a <code>DataFrame</code>.</p></dd>


<dt id="arg-bs-genome">bs_genome<a class="anchor" aria-label="anchor" href="#arg-bs-genome"></a></dt>
<dd><p>A character value that indicates the reference genome name as
defined by <code>BSgenome</code> package. Default is 'BSgenome.Hsapiens.UCSC.hg19'.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed into <code><a href="https://rdrr.io/pkg/MEDIPS/man/MEDIPS.createROIset.html" class="external-link">MEDIPS::MEDIPS.createROIset</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An object of class <code>SummarizedExperiment</code> with read count matrix stored
as an assay named 'counts' (can be extracted using <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment::assays</a></code>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">getRoiReadCount</span><span class="op">(</span></span></span>
<span class="r-in"><span> sample_bam_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'dir/to/bam1'</span>,<span class="st">'dir/to/bam2'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> sample_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'sample1'</span>, <span class="st">'sample2'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span> sample_paired <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in getRoiReadCount(sample_bam_files = c("dir/to/bam1", "dir/to/bam2"),     sample_names = c("sample1", "sample2"), sample_paired = TRUE):</span> argument "roi" is missing, with no default</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ning Shen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

