<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Cell type deconvolutiond of (cf)MeDIP-seq data with decemedip ‚Ä¢ decemedip</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Cell type deconvolutiond of (cf)MeDIP-seq data with decemedip">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">decemedip</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/how-to-use-decemedip.html">Cell type deconvolutiond of (cf)MeDIP-seq data with decemedip</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Cell type deconvolutiond of (cf)MeDIP-seq data with decemedip</h1>
                        <h4 data-toc-skip class="author">Ning Shen</h4>
            <address class="author_afil">
      Department of Statistics, University of British Columbia;Centre
for Molecular Medicine and Therapeutics, BC Children‚Äôs Hospital Research
Institute<br><a class="author_email" href="mailto:#"></a><a href="mailto:ning.shen.wk@gmail.com" class="email">ning.shen.wk@gmail.com</a>
      </address>
                  
      

      <div class="d-none name"><code>how-to-use-decemedip.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p><strong>If you use decemedip in published research, please cite:
</strong></p>
<p>Shen, Ning, and Keegan Korthauer. ‚Äúdecemedip: hierarchical Bayesian
modeling for cell type deconvolution of immunoprecipitation-based DNA
methylome‚Äù bioRxiv, forthcoming.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>This package will be submitted to Bioconductor. For now, you can
install the development version from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install stable version from Bioconductor (once available)</span></span>
<span><span class="co"># BiocManager::install("decemedip")</span></span>
<span></span>
<span><span class="co"># Install development version from GitHub</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"nshen7/decemedip"</span><span class="op">)</span></span></code></pre></div>
<p>After installation, load the decemedip package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">decemedip</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>Cell-free and bulk DNA methylation data obtained through MeDIP-seq
reflect a mixture of methylation signals across multiple cell types.
Decomposing these signals to infer cell type composition can provide
valuable insights for cancer diagnosis, immune response monitoring, and
other biomedical applications. However, challenges like
enrichment-induced biases and sparse reference data make this task
complex. The <code>decemedip</code> package addresses these challenges
through a hierarchical Bayesian framework that estimates cell type
proportions and models the relationship between MeDIP-seq counts and
reference methylation data.</p>
<p>decemedip couples a logit-normal model with a generalize additive
model (GAM) framework. For each site
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>‚àà</mo><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mi>.</mi><mi>.</mi><mi>.</mi><mo>,</mo><mi>N</mi><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">i \in \{1, ..., N\}</annotation></semantics></math>
in the reference panel, the input to the model is the fractional
methylation levels
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><mrow><mi>i</mi><mi>k</mi></mrow></msub><annotation encoding="application/x-tex">x_{ik}</annotation></semantics></math>
for each cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mo>‚àà</mo><mo stretchy="false" form="prefix">{</mo><mn>1</mn><mo>,</mo><mi>.</mi><mi>.</mi><mi>.</mi><mo>,</mo><mi>K</mi><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">k \in \{1, ..., K\}</annotation></semantics></math>,
the CpG density level
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>z</mi><mi>i</mi></msub><annotation encoding="application/x-tex">z_i</annotation></semantics></math>
and the MeDIP-seq read count
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>y</mi><mi>i</mi></msub><annotation encoding="application/x-tex">y_i</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
is the total number of cell types. A unit simplex variable that follows
a logit-normal prior,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ùõë</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>œÄ</mi><mn>1</mn></msub><mo>,</mo><mi>.</mi><mi>.</mi><mi>.</mi><mo>,</mo><msub><mi>œÄ</mi><mi>K</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\boldsymbol{\pi} = (\pi_1, ..., \pi_K)</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>œÄ</mi><mi>k</mi></msub><mo>&gt;</mo><mn>0</mn><mo>,</mo><msubsup><mo>‚àë</mo><mrow><mi>k</mi><mo>=</mo><mn>1</mn></mrow><mi>K</mi></msubsup><msub><mi>œÄ</mi><mi>k</mi></msub><mo>=</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">\pi_k &gt; 0, \sum_{k=1}^K \pi_k = 1</annotation></semantics></math>,
is included to describe the proportions of cell types in the reference
panel while taking into account the correlations between these cell
types.</p>
<p><img src="../reference/figures/method_main_figure.png" width="640" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="input-data">Input Data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p><code>decemedip</code> requires three primary inputs:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Reference methylation matrix</strong>: A matrix of
methylation levels for multiple cell types at selected CpG sites.</li>
<li>
<strong>CpG density</strong>: Information about CpG site density in
the genome.</li>
<li>
<strong>MeDIP-seq read counts</strong>: Coverage values from the
sample of interest.</li>
</ol>
<div class="section level3">
<h3 id="reference-panel">Reference panel<a class="anchor" aria-label="anchor" href="#reference-panel"></a>
</h3>
<p>Our package provides default reference matrices for hg19 and hg38
along with the corresponding CpG density information, as objects in
class <code>SummarizedExperiment</code>. The objects can be accessed by
calling <code>data(hg19.ref.cts.se)</code> and
<code>data(hg19.ref.anc.se)</code>, or
<code>data(hg38.ref.cts.se)</code> and
<code>data(hg38.ref.anc.se)</code>. By default, the main function
<code><a href="../reference/decemedip.html">decemedip()</a></code> applies <code>hg19.ref.cts.se</code> and
<code>hg19.ref.anc.se</code> as the reference panels. Please refer to
the manuscript for details of how the default reference panels were
constructed.</p>
<p>On a side note, we provide the function
<code><a href="../reference/makeReferencePanel.html">makeReferencePanel()</a></code> to allow user to build their own
reference panels, which only requires input of reference CpGs and
corresponded fractional methylation level matrix. The function computes
CpG density on its own. Note that the cell type-specific sites and
anchor sites need to be included in two
<code>SummarizedExperiment</code> objects to be inputs to the main
function <code><a href="../reference/decemedip.html">decemedip()</a></code>. See <code><a href="../reference/makeReferencePanel.html">?makeReferencePanel</a></code>
for more information.</p>
<p>We show how the reference panel look like in the following
chunks:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hg19.ref.cts.se</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">hg19.ref.cts.se</span><span class="op">)</span></span>
<span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 2500 25 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): beta</span></span>
<span><span class="co">## rownames(2500): cg18856478 cg20820767 ... cg01071459 cg20726993</span></span>
<span><span class="co">## rowData names(4): probe label pos n_cpgs_100bp</span></span>
<span><span class="co">## colnames(25): Monocytes_EPIC B-cells_EPIC ... Upper_GI Uterus_cervix</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">hg19.ref.cts.se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## GRanges object with 6 ranges and 4 metadata columns:</span></span>
<span><span class="co">##              seqnames    ranges strand |       probe                  label</span></span>
<span><span class="co">##                 &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">##   cg18856478     chr1  43814358      * |  cg18856478 Monocytes_EPIC hyper..</span></span>
<span><span class="co">##   cg20820767     chr1  45082840      * |  cg20820767 Monocytes_EPIC hyper..</span></span>
<span><span class="co">##   cg14855367     chr3 191048308      * |  cg14855367 Monocytes_EPIC hyper..</span></span>
<span><span class="co">##   cg25913761    chr15  90727560      * |  cg25913761 Monocytes_EPIC hyper..</span></span>
<span><span class="co">##   cg21546950     chr1  77904032      * |  cg21546950 Monocytes_EPIC hyper..</span></span>
<span><span class="co">##   cg14981189    chr10   5113871      * |  cg14981189 Monocytes_EPIC hyper..</span></span>
<span><span class="co">##                    pos n_cpgs_100bp</span></span>
<span><span class="co">##              &lt;integer&gt;    &lt;integer&gt;</span></span>
<span><span class="co">##   cg18856478  43814358            7</span></span>
<span><span class="co">##   cg20820767  45082840           12</span></span>
<span><span class="co">##   cg14855367 191048308            1</span></span>
<span><span class="co">##   cg25913761  90727560            4</span></span>
<span><span class="co">##   cg21546950  77904032            1</span></span>
<span><span class="co">##   cg14981189   5113871            1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 22 sequences from hg19 genome; no seqlengths</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hg19.ref.anc.se</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">hg19.ref.anc.se</span><span class="op">)</span></span>
<span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 1000 25 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): beta</span></span>
<span><span class="co">## rownames(1000): 353534 76294 ... 73948 87963</span></span>
<span><span class="co">## rowData names(7): probe pos ... avg_beta_rank n_cpgs_100bp</span></span>
<span><span class="co">## colnames(25): Monocytes_EPIC B-cells_EPIC ... Upper_GI Uterus_cervix</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">hg19.ref.anc.se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## GRanges object with 6 ranges and 7 metadata columns:</span></span>
<span><span class="co">##          seqnames    ranges strand |       probe       pos        label</span></span>
<span><span class="co">##             &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;  &lt;character&gt;</span></span>
<span><span class="co">##   353534    chr19  19496477      * |  cg15264323  19496477 All-tissue U</span></span>
<span><span class="co">##    76294     chr1 154193656      * |  cg13576006 154193656 All-tissue U</span></span>
<span><span class="co">##   300825    chr12    860787      * |  cg13284045    860787 All-tissue U</span></span>
<span><span class="co">##   369864     chr5  43514988      * |  cg20545087  43514988 All-tissue U</span></span>
<span><span class="co">##   247131    chr16    278788      * |  cg06819375    278788 All-tissue U</span></span>
<span><span class="co">##   308091    chr12  96429111      * |  cg09725090  96429111 All-tissue U</span></span>
<span><span class="co">##             margin  avg_beta avg_beta_rank n_cpgs_100bp</span></span>
<span><span class="co">##          &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;    &lt;integer&gt;</span></span>
<span><span class="co">##   353534       0.1  0.063240       70720.0           10</span></span>
<span><span class="co">##    76294       0.1  0.013492       19877.5            7</span></span>
<span><span class="co">##   300825       0.1  0.029155       36307.5            1</span></span>
<span><span class="co">##   369864       0.1  0.038771       45387.5            6</span></span>
<span><span class="co">##   247131       0.1  0.032970       39748.5            2</span></span>
<span><span class="co">##   308091       0.1  0.076555       77269.0            7</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 22 sequences from hg19 genome; no seqlengths</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="fit-the-bayesian-model">Fit the Bayesian Model<a class="anchor" aria-label="anchor" href="#fit-the-bayesian-model"></a>
</h2>
<p>The main function <code><a href="../reference/decemedip.html">decemedip()</a></code> fits the decemedip model.
It allows two types of input:</p>
<ol style="list-style-type: decimal">
<li>A BAM file of the sample MeDIP-seq data, or</li>
<li>read counts of the reference CpG sites of the sample.</li>
</ol>
<p>We provide instructions for both input types as follows.</p>
<div class="section level3">
<h3 id="input-with-a-bam-file">Input with a BAM file<a class="anchor" aria-label="anchor" href="#input-with-a-bam-file"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_bam_file</span> <span class="op">&lt;-</span> <span class="st">"path/to/bam/files"</span></span>
<span><span class="va">paired</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span> <span class="co"># whether the sequencing is paired-end</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/decemedip.html">decemedip</a></span><span class="op">(</span>sample_bam_file <span class="op">=</span> <span class="va">sample_bam_file</span>, </span>
<span>                    paired <span class="op">=</span> <span class="va">paired</span>, </span>
<span>                    cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p>PS: By default, the <code><a href="../reference/decemedip.html">decemedip()</a></code> function uses a hg19
reference panel. But users may add the arguments
<code>ref_cts = hg38.ref.cts.se, ref_anc = hg38.ref.anc.se</code> to
apply read counts extraction on hg38 data.</p>
</div>
<div class="section level3">
<h3 id="input-with-read-counts">Input with read counts<a class="anchor" aria-label="anchor" href="#input-with-read-counts"></a>
</h3>
<p>We use built-in objects of the package that contains read counts of
the prostate tumor patient-derived xenograft (PDX) samples from the
Berchuck et al.¬†2022 [1] study, <code>pdx.counts.cts.se</code> and
<code>pdx.counts.anc.se</code>, to demonstrate the output and
diagnostics in this vignette.</p>
<p>[1] Berchuck JE, Baca SC, McClure HM, Korthauer K, Tsai HK, Nuzzo PV,
et al.¬†Detecting neuroendocrine prostate cancer through tissue-informed
cell-free DNA methylation analysis. Clinical Cancer Research.
2022;28(5):928‚Äì938.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pdx.counts.cts.se</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pdx.counts.anc.se</span><span class="op">)</span></span></code></pre></div>
<p>We extract the sample <code>LuCaP_147CR</code> from this example
dataset for follwing illustration:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_cts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">pdx.counts.cts.se</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>,<span class="st">'LuCaP_147CR'</span><span class="op">]</span> <span class="co"># read counts of cell type-specific CpGs of the sample 'LuCaP_147CR'</span></span>
<span><span class="va">counts_anc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">pdx.counts.anc.se</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>,<span class="st">'LuCaP_147CR'</span><span class="op">]</span> <span class="co"># read counts of anchor CpGs of the sample 'LuCaP_147CR'</span></span></code></pre></div>
<p>Due to the vignette running time limit by Bioconductor, we only run
300 iterations (<code>iter = 300</code>) for the purpose of
demonstration, which causes the warning of
<code>Bulk Effective Samples Size (ESS) is too low</code>. In regular
cases, we recommend to run 2000 iterations (the default), or at least
1000 iterations for a stable posterior inference.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/decemedip.html">decemedip</a></span><span class="op">(</span>counts_cts <span class="op">=</span> <span class="va">counts_cts</span>,</span>
<span>                    counts_anc <span class="op">=</span> <span class="va">counts_anc</span>,</span>
<span>                    diagnostics <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    cores <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                    iter <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="co">## MCMC converged with seed 2024</span></span></code></pre></div>
<p>The output of <code><a href="../reference/decemedip.html">decemedip()</a></code> is a list containing two
elements:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">output</span><span class="op">)</span></span>
<span><span class="co">## [1] "data_list" "posterior"</span></span></code></pre></div>
<ul>
<li>
<code>data_list</code>: An organized list of variables used as input
to the Stan posterior sampling function.</li>
<li>
<code>posterior</code>: An <code>stanfit</code> object produced by
Stan representing the fitted posteriors.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="checking-model-outputs">Checking model outputs<a class="anchor" aria-label="anchor" href="#checking-model-outputs"></a>
</h2>
<div class="section level3">
<h3 id="cell-type-proportions">Cell type proportions<a class="anchor" aria-label="anchor" href="#cell-type-proportions"></a>
</h3>
<p>After running the model, you may extract and save the summary of
fitted posteriors using the <code><a href="https://mc-stan.org/rstan/reference/monitor.html" class="external-link">monitor()</a></code> and
<code><a href="https://mc-stan.org/rstan/reference/stanfit-method-extract.html" class="external-link">extract()</a></code> functions provided by the <code>RStan</code>
package. See documentation of <code>RStan</code> for details of these
functions.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mc-stan.org/rstan/" class="external-link">rstan</a></span><span class="op">)</span></span></code></pre></div>
<p>Extract the fitted posterior of cell type proportions
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>ùõë</mi><annotation encoding="application/x-tex">\boldsymbol\pi</annotation></semantics></math>):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">smr_pi.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSummaryOnPi.html">getSummaryOnPi</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">posterior</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">smr_pi.df</span><span class="op">)</span></span>
<span><span class="co">##              cell_type    mean se_mean      sd    2.5%     25%     50%     75%</span></span>
<span><span class="co">## pi[1]   Monocytes_EPIC 0.00041 9.9e-05 0.00136 8.7e-14 1.2e-07 2.3e-06 7.4e-05</span></span>
<span><span class="co">## pi[2]     B-cells_EPIC 0.00021 4.6e-05 0.00055 2.7e-13 5.4e-08 4.0e-06 1.0e-04</span></span>
<span><span class="co">## pi[3]  CD4T-cells_EPIC 0.00042 1.2e-04 0.00143 2.4e-13 9.9e-08 4.6e-06 8.1e-05</span></span>
<span><span class="co">## pi[4]    NK-cells_EPIC 0.00040 1.9e-04 0.00262 7.0e-14 2.7e-08 1.4e-06 3.0e-05</span></span>
<span><span class="co">## pi[5]  CD8T-cells_EPIC 0.00049 1.8e-04 0.00264 2.3e-13 8.7e-08 2.2e-06 4.1e-05</span></span>
<span><span class="co">## pi[6] Neutrophils_EPIC 0.00035 1.0e-04 0.00144 1.0e-13 2.8e-08 2.2e-06 5.0e-05</span></span>
<span><span class="co">##        97.5% n_eff Rhat valid</span></span>
<span><span class="co">## pi[1] 0.0043   162    1     1</span></span>
<span><span class="co">## pi[2] 0.0017   139    1     1</span></span>
<span><span class="co">## pi[3] 0.0050   142    1     1</span></span>
<span><span class="co">## pi[4] 0.0031   175    1     1</span></span>
<span><span class="co">## pi[5] 0.0049   215    1     1</span></span>
<span><span class="co">## pi[6] 0.0044   205    1     1</span></span></code></pre></div>
<ul>
<li>Summary Statistics Columns:
<ul>
<li>
<code>cell_type</code>: The name of the parameter or variable being
analyzed.</li>
<li>
<code>mean</code>: The posterior mean, representing the point
estimate of the parameter.</li>
<li>
<code>se_mean</code>: The standard error of the mean, calculated as
sd / sqrt(n_eff), indicating precision of the mean estimate.</li>
<li>
<code>sd</code>: The posterior standard deviation, representing the
spread or uncertainty of the parameter estimate.</li>
<li>
<code>2.5%</code>, <code>25%</code>, <code>50%</code> (median),
<code>75%</code>, <code>97.5%</code>: Percentiles of the posterior
distribution, providing a summary of parameter uncertainty. These define
the 95% credible interval (2.5% to 97.5%).</li>
</ul>
</li>
<li>Diagnostics Columns:
<ul>
<li>
<code>n_eff</code>: The effective sample size, indicating how many
independent samples the chain produced after accounting for
autocorrelation.</li>
<li>
<code>Rhat</code>: The potential scale reduction factor, measuring
chain convergence. Values close to 1.00 suggest good convergence.</li>
<li>
<code>valid</code>: A flag indicating whether diagnostic checks
(e.g., Rhat and n_eff) passed for this parameter (1 = passed, 0 =
potential issues).</li>
</ul>
</li>
</ul>
<p>Plotting out the fitted cell type proportions with credible
intervals:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'_'</span>, <span class="st">' '</span>, <span class="va">smr_pi.df</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">'(.*) EPIC'</span>, <span class="st">'\\1'</span>, <span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">smr_pi.df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>cell_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cell_type</span>, labels <span class="op">=</span> <span class="va">labels</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">cell_type</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_linerange</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">`2.5%`</span>, ymax <span class="op">=</span> <span class="va">`97.5%`</span><span class="op">)</span>, </span>
<span>                 position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge2</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.035</span><span class="op">)</span>,</span>
<span>                 linewidth <span class="op">=</span> <span class="fl">7</span>, alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_linerange</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>ymin <span class="op">=</span> <span class="va">`25%`</span>, ymax <span class="op">=</span> <span class="va">`75%`</span><span class="op">)</span>, </span>
<span>                 position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge2</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.035</span><span class="op">)</span>,</span>
<span>                 linewidth <span class="op">=</span> <span class="fl">7</span>, alpha <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge2</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.035</span><span class="op">)</span>,</span>
<span>             fill <span class="op">=</span> <span class="st">'white'</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="how-to-use-decemedip_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="fitted-relationship-between-fractional-methylation-and-medip-seq-counts">Fitted relationship between fractional methylation and MeDIP-seq
counts<a class="anchor" aria-label="anchor" href="#fitted-relationship-between-fractional-methylation-and-medip-seq-counts"></a>
</h3>
<p>Note that this plot is only accessible when <code>diagnostics</code>
is set to<code>TRUE</code> in the <code><a href="../reference/decemedip.html">decemedip()</a></code> function. The
actual read counts (orange) and the fitted counts predicted by the GAM
component (black) are shown in the figure across varying levels of CpG
density. Grey area represents the 95% credible intervals of the
predicted counts. `CpG density: x‚Äô means that there are x CpGs in the
100-bp window surrounding the CpG.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDiagnostics.html">plotDiagnostics</a></span><span class="op">(</span><span class="va">output</span>, plot_type <span class="op">=</span> <span class="st">"y_fit"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">300</span><span class="op">)</span></span></code></pre></div>
<p><img src="how-to-use-decemedip_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>The <code>decemedip</code> package provides a robust framework for
cell type deconvolution from MeDIP-seq data. By following this vignette,
users can apply the method to their own datasets, extract key model
outputs, and generate diagnostic plots for analysis.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] rstan_2.32.7                StanHeaders_2.32.10        </span></span>
<span><span class="co">##  [3] decemedip_0.99.1            testthat_3.2.3             </span></span>
<span><span class="co">##  [5] ggplot2_3.5.2               dplyr_1.1.4                </span></span>
<span><span class="co">##  [7] SummarizedExperiment_1.38.1 Biobase_2.68.0             </span></span>
<span><span class="co">##  [9] GenomicRanges_1.60.0        GenomeInfoDb_1.44.0        </span></span>
<span><span class="co">## [11] IRanges_2.42.0              S4Vectors_0.46.0           </span></span>
<span><span class="co">## [13] BiocGenerics_0.54.0         generics_0.1.4             </span></span>
<span><span class="co">## [15] MatrixGenerics_1.20.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [17] BiocStyle_2.36.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       rstudioapi_0.17.1        jsonlite_2.0.0          </span></span>
<span><span class="co">##   [4] magrittr_2.0.3           farver_2.1.2             rmarkdown_2.29          </span></span>
<span><span class="co">##   [7] fs_1.6.6                 BiocIO_1.18.0            ragg_1.4.0              </span></span>
<span><span class="co">##  [10] vctrs_0.6.5              memoise_2.0.1            Rsamtools_2.24.0        </span></span>
<span><span class="co">##  [13] RCurl_1.98-1.17          usethis_3.1.0            htmltools_0.5.8.1       </span></span>
<span><span class="co">##  [16] S4Arrays_1.8.0           progress_1.2.3           curl_6.2.2              </span></span>
<span><span class="co">##  [19] SparseArray_1.8.0        sass_0.4.10              bslib_0.9.0             </span></span>
<span><span class="co">##  [22] htmlwidgets_1.6.4        desc_1.4.3               httr2_1.1.2             </span></span>
<span><span class="co">##  [25] cachem_1.1.0             GenomicAlignments_1.44.0 mime_0.13               </span></span>
<span><span class="co">##  [28] lifecycle_1.0.4          pkgconfig_2.0.3          Matrix_1.7-3            </span></span>
<span><span class="co">##  [31] R6_2.6.1                 fastmap_1.2.0            GenomeInfoDbData_1.2.14 </span></span>
<span><span class="co">##  [34] shiny_1.10.0             digest_0.6.37            ps_1.9.1                </span></span>
<span><span class="co">##  [37] AnnotationDbi_1.70.0     rprojroot_2.0.4          pkgload_1.4.0           </span></span>
<span><span class="co">##  [40] textshaping_1.0.1        RSQLite_2.3.11           labeling_0.4.3          </span></span>
<span><span class="co">##  [43] filelock_1.0.3           MEDIPS_1.60.0            httr_1.4.7              </span></span>
<span><span class="co">##  [46] abind_1.4-8              compiler_4.5.0           remotes_2.5.0           </span></span>
<span><span class="co">##  [49] bit64_4.6.0-1            withr_3.0.2              inline_0.3.21           </span></span>
<span><span class="co">##  [52] BiocParallel_1.42.0      DBI_1.2.3                QuickJSR_1.7.0          </span></span>
<span><span class="co">##  [55] pkgbuild_1.4.7           R.utils_2.13.0           biomaRt_2.64.0          </span></span>
<span><span class="co">##  [58] sessioninfo_1.2.3        rappdirs_0.3.3           DelayedArray_0.34.1     </span></span>
<span><span class="co">##  [61] rjson_0.2.23             gtools_3.9.5             DNAcopy_1.82.0          </span></span>
<span><span class="co">##  [64] loo_2.8.0                BH_1.87.0-1              tools_4.5.0             </span></span>
<span><span class="co">##  [67] httpuv_1.6.16            R.oo_1.27.1              glue_1.8.0              </span></span>
<span><span class="co">##  [70] callr_3.7.6              restfulr_0.0.15          promises_1.3.2          </span></span>
<span><span class="co">##  [73] grid_4.5.0               gtable_0.3.6             BSgenome_1.76.0         </span></span>
<span><span class="co">##  [76] R.methodsS3_1.8.2        preprocessCore_1.70.0    hms_1.1.3               </span></span>
<span><span class="co">##  [79] xml2_1.3.8               XVector_0.48.0           pillar_1.10.2           </span></span>
<span><span class="co">##  [82] stringr_1.5.1            later_1.4.2              BiocFileCache_2.16.0    </span></span>
<span><span class="co">##  [85] lattice_0.22-6           rtracklayer_1.68.0       bit_4.6.0               </span></span>
<span><span class="co">##  [88] tidyselect_1.2.1         Biostrings_2.76.0        miniUI_0.1.2            </span></span>
<span><span class="co">##  [91] knitr_1.50               gridExtra_2.3            bookdown_0.43           </span></span>
<span><span class="co">##  [94] xfun_0.52                brio_1.1.5               devtools_2.4.5          </span></span>
<span><span class="co">##  [97] stringi_1.8.7            UCSC.utils_1.4.0         yaml_2.3.10             </span></span>
<span><span class="co">## [100] evaluate_1.0.3           codetools_0.2-20         RcppEigen_0.3.4.0.2     </span></span>
<span><span class="co">## [103] tibble_3.2.1             BiocManager_1.30.25      cli_3.6.5               </span></span>
<span><span class="co">## [106] RcppParallel_5.1.10      xtable_1.8-4             systemfonts_1.2.3       </span></span>
<span><span class="co">## [109] processx_3.8.6           jquerylib_0.1.4          Rcpp_1.0.14             </span></span>
<span><span class="co">## [112] dbplyr_2.5.0             png_0.1-8                XML_3.99-0.18           </span></span>
<span><span class="co">## [115] parallel_4.5.0           rstantools_2.4.0         ellipsis_0.3.2          </span></span>
<span><span class="co">## [118] pkgdown_2.1.2            blob_1.2.4               prettyunits_1.2.0       </span></span>
<span><span class="co">## [121] profvis_0.4.0            bayesplot_1.12.0         urlchecker_1.0.1        </span></span>
<span><span class="co">## [124] bitops_1.0-9             scales_1.4.0             purrr_1.0.4             </span></span>
<span><span class="co">## [127] crayon_1.5.3             rlang_1.1.6              cowplot_1.1.3           </span></span>
<span><span class="co">## [130] KEGGREST_1.48.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ning Shen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
